### ISSUE TITLE
Fix and enable Apache 2.2 support

### ISSUE THREAD

**Issue by pde** [letsencrypt/letsencrypt#2228]
Inheriting the original intent of #2012.

**Comment by SwartzCr** [comment]
on apache 2.2 there seems to be some problems with writing and reverting various config files
for example apache2.conf never gets the include line added that specifies the le_tls_sni_01_cert_challenge.conf file
additionally when attempting to restart apache we get this error
Warning: DocumentRoot [/var/lib/letsencrypt/tls_sni_01_page/] does not exist
apache2: Could not reliably determine the server's fully qualified domain name, using 127.0.1.1 for ServerName
[Wed Jan 20 08:28:00 2016] [warn] NameVirtualHost *:80 has no VirtualHosts
[Wed Jan 20 08:28:00 2016] [error] Server should be SSL-aware but has no certificate configured [Hint: SSLCertificateFile] ((null):0)
[Wed Jan 20 08:51:08 2016] [error] Server should be SSL-aware but has no certificate configured [Hint: SSLCertificateFile] ((null):0)
[Wed Jan 20 08:52:09 2016] [warn] pid file /var/run/apache2.pid overwritten -- Unclean shutdown of previous Apache run?
[Wed Jan 20 08:52:09 2016] [notice] Apache/2.2.22 (Ubuntu) configured -- resuming normal operations

**Comment by SwartzCr** [comment]
oops that was my broken configuration. Adding and removing the challenge include solves the problem and allows a cert to be fetched

**Comment by SwartzCr** [comment]
augeas also doesn't seem to notice that it's made this change, if we call the initial:
```143         if len(self.configurator.parser.find_dir(                                
144                 parser.case_i("Include"), self.challenge_conf)) == 0:```
as in `_conf_include_check` after the `self.configurator.parser.add_dir` call we again get a len of 0

I am pretty confident it is trying to set the path right, when I include these debug lines:
```145             logger.debug("Including challenge virtual host(s)")                  
146             logger.debug("main config is %s" % (main_config))                    
147             logger.debug("aug_path is %s, challenge conf is %s" % (parser.get_aug_path(main_config), self.challenge_conf))
148             self.configurator.parser.add_dir(                                    
149                 parser.get_aug_path(main_config),                                
150                 "Include", self.challenge_conf)      ```

I get back:
```2016-01-20 19:23:15,710:DEBUG:letsencrypt_apache.tls_sni_01:Including challenge virtual host(s)
2016-01-20 19:23:15,711:DEBUG:letsencrypt_apache.tls_sni_01:main config is /etc/apache2/apache2.conf
2016-01-20 19:23:15,712:DEBUG:letsencrypt_apache.tls_sni_01:aug_path is /files/etc/apache2/apache2.conf, challenge conf is /etc/apache2/le_tls_sni_01_cert_challenge.conf```

**Comment by SwartzCr** [comment]
SO the problem seems to be that self.aug.match throws out unsaved changes:
```
ipdb> self.aug.get("/etc/apache2/apache2.conf/directive[25]")
ipdb> self.aug.get("/files/etc/apache2/apache2.conf/directive[25]")
u'Include'
ipdb> self.aug.get("/files/etc/apache2/apache2.conf/directive[25]/arg")
u'/etc/apache2/le_tls_sni_01_cert_challenge.conf'
ipdb> self._get_include_path(self.get_arg(match + "/arg")
*** SyntaxError: unexpected EOF while parsing (<stdin>, line 1)
ipdb> self._get_include_path(self.get_arg(match + "/arg"))
*** NameError: name 'match' is not defined
ipdb> match = matches[-1]
ipdb> match
u'/files/etc/apache2/apache2.conf/directive[25]'
ipdb> self._get_include_path(self.get_arg(match + "/arg"))
u'/files/etc/apache2/le_tls_sni_01_cert_challenge.conf'
ipdb> self.aug.match(match)
[]
ipdb> self.aug.match(match+arg_suffix)
[]
ipdb> self.aug.get("/files/etc/apache2/apache2.conf/directive[25]/arg")
ipdb> self.aug.get("/files/etc/apache2/apache2.conf/directive[24]/arg")
u'sites-enabled/'
```
current believed culprit is https://github.com/hercules-team/python-augeas/blob/master/augeas.py#L432

**Comment by SwartzCr** [comment]
adding self.aug.save() to the end of parser.add_dir(...) fixes the initial issue

**Comment by SwartzCr** [comment]
```
213     def _get_ifmod(self, aug_conf_path, mod):                                    
214         """Returns the path to <IfMod mod> and creates one if it doesn't exist.  
215                                                                                  
216         :param str aug_conf_path: Augeas configuration path                      
217         :param str mod: module ie. mod_ssl.c                                     
218                                                                                  
219         """                                                                      
220         if_mods = self.aug.match(("%s/IfModule/*[self::arg='%s']" %              
221                                   (aug_conf_path, mod)))    
```
maybe?

**Comment by SwartzCr** [comment]
```
652     def make_addrs_sni_ready(self, addrs):                                       
 653         """Checks to see if the server is ready for SNI challenges.              
 654                                                                                  
 655         :param addrs: Addresses to check SNI compatibility                       
 656         :type addrs: :class:`~letsencrypt_apache.obj.Addr`                       
 657                                                                                  
 658         """                                                                      
 659         # Version 2.4 and later are automatically SNI ready.                     
 660         if self.version >= (2, 4):                                               
 661             return                                                               
 662                                                                                  
 663         for addr in addrs:                                                       
 664             if not self.is_name_vhost(addr):                                     
 665                 logger.debug("Setting VirtualHost at %s to be a name "           
 666                              "based virtual host", addr)                         
 667                 self.add_name_vhost(addr)     
```
calls
```
 563     def add_name_vhost(self, addr):                                              
 564         """Adds NameVirtualHost directive for given address.                     
 565                                                                                  
 566         :param addr: Address that will be added as NameVirtualHost directive     
 567         :type addr: :class:`~letsencrypt_apache.obj.Addr`                        
 568                                                                                  
 569         """                                                                      
 570         loc = parser.get_aug_path(self.parser.loc["name"])                       
 571         logger.debug("adding ssl to file at path %s" % (loc))                    
 572         if addr.get_port() == "443":                                             
 573             path = self.parser.add_dir_to_ifmodssl(                              
 574                 loc, "NameVirtualHost", [str(addr)])  
```
calls
```
187     def add_dir_to_ifmodssl(self, aug_conf_path, directive, args):               
188         """Adds directive and value to IfMod ssl block.                          
189                                                                                  
190         Adds given directive and value along configuration path within           
191         an IfMod mod_ssl.c block.  If the IfMod block does not exist in          
192         the file, it is created.                                                 
193                                                                                  
194         :param str aug_conf_path: Desired Augeas config path to add directive    
195         :param str directive: Directive you would like to add, e.g. Listen       
196         :param args: Values of the directive; str "443" or list of str           
197         :type args: list                                                         
198                                                                                  
199         """                                                                      
200         # TODO: Add error checking code... does the path given even exist?       
201         #       Does it throw exceptions?                                        
202         if_mod_path = self._get_ifmod(aug_conf_path, "mod_ssl.c")         
```
calls
```
213     def _get_ifmod(self, aug_conf_path, mod):                                    
214         """Returns the path to <IfMod mod> and creates one if it doesn't exist.  
215                                                                                  
216         :param str aug_conf_path: Augeas configuration path                      
217         :param str mod: module ie. mod_ssl.c                                     
218                                                                                  
219         """                                                                      
220         if_mods = self.aug.match(("%s/IfModule/*[self::arg='%s']" %              
221                                   (aug_conf_path, mod)))  
```
and the `self.aug.match(...)` gets rid of our unsaved changes, which include the addition of the directive for Include le_challenge_whatever.conf in apache2.conf

**Comment by SwartzCr** [comment]
jfc
```
  /root/letsencrypt/letsencrypt-apache/letsencrypt_apache/configurator.py(561)is_name_vhost()
    559         # note ip_addr can be FQDN although Apache does not recommend it
    560         return (self.version >= (2, 4) or
--> 561                 self.parser.find_dir("NameVirtualHost", str(target_addr)))
    562 
    563     def add_name_vhost(self, addr):

> /root/letsencrypt/letsencrypt-apache/letsencrypt_apache/parser.py(284)find_dir()
    282         # Cannot place member variable in the definition of the function so...
    283         ipdb.set_trace()
--> 284         if not start:
    285             start = get_aug_path(self.loc["root"])
    286 
```

**Comment by SwartzCr** [comment]
```
ipdb> self.aug.get("/files/etc/apache2/apache2.conf/directive[last()]/arg")
u'/etc/apache2/le_tls_sni_01_cert_challenge.conf'
ipdb> c
> /root/letsencrypt/letsencrypt-apache/letsencrypt_apache/parser.py(284)find_dir()
    283         ipdb.set_trace()
--> 284         if not start:
    285             start = get_aug_path(self.loc["root"])

ipdb> self.aug.get("/files/etc/apache2/apache2.conf/directive[last()]/arg")
u'sites-enabled/'
ipdb> 
```

**Comment by SwartzCr** [comment]
Bug: self.aug.match() should not be called before saving changes made by augeas

**Comment by bmw** [comment]
Reopening. The remaining problem that I'm aware of are `_default_:443` and `*;443` vhosts interacting differently in Apache 2.2 than in Apache 2.4, causing us to setup conflicting vhosts.

**Comment by pde** [comment]
Conclusion for 0.4.0: we see if #1991 basically fixes Apache 2.2 for us; if yes, it ships in 0.4.0.  If not, we're kicking to 0.5.0.

**Comment by bmw** [comment]
Fixed in #2286.
